### PR TITLE
Retry NAT punching without duration on mapping failure

### PR DESCRIPTION
**Pull request by Kubuxu**
Some hardware doesn't support UPnP with durations.

### CODE REVIEW

**Comment by Kubuxu** [comment]
//c @whyrusleeping for review

**Comment by whyrusleeping** [comment]
what happens if we try to set a timeout on a router that doesnt support timeouts?

**Comment by Kubuxu** [comment]
The request fails, no mapping is created, this means that we didn't punch the NAT.

**Comment by whyrusleeping** [comment]
If a request fails, can we tell that its because timeouts are not supported and not for some other reason?

**Comment by lgierth** [comment]
Oh lol I can see that happening :) Apart from this it's good to regularly retry the mapping even if it succeeded in the first place -- routers reset, forget stuff, etc.

Rule of thumb: assume that your router is crap in one way or another until proven otherwise :)

**Comment by lgierth** [comment]
> Apart from this it's good to regularly retry the mapping even if it succeeded in the first place

But that's for a different PR -- LGTM!

**Comment by Kubuxu** [comment]
@whyrusleeping we can't, go nat gives us only generic error that SOAP call failed.

@lgierth didn't think about that, I will add it, shouldn't be hard.

**Comment by whyrusleeping** [comment]
Yeah, this LGTM. only thing i would prefer is matching the errors more closely, but i doubt that we would get a consistent set of errors back in all cases

**Comment by Kubuxu** [comment]
Hmm, it looks like calling for new permanent mapping on the same port with the same agent returns same mapping.

So I can just not short circuit calls in case of permanent mappings.

@whyrusleeping retrying it with no timeout doesn't cause any damage and has just a benefit, worst case it will fail again, best case it will work.

**Comment by whyrusleeping** [comment]
Yeah, that seems like a good idea to me

**Comment by Kubuxu** [comment]
Updated.

**Comment by jbenet** [comment]
> it looks like calling for new permanent mapping on the same port with the same agent returns same mapping.

For some NATs. Some NATs likely don't do that. Few things are always true in networking but "there is a middlebox that will do that" is a close one.

Basically, try and try again in a variety of ways, and you'll succeed.

**Comment by Kubuxu** [comment]
@lgierth @whyrusleeping does it look good now?

File: p2p/nat/nat.go [hunk 1]
```diff
@@ -333,13 +336,23 @@ func (nat *NAT) NewMapping(maddr ma.Multiaddr) (Mapping, error) {
 
 func (nat *NAT) establishMapping(m *mapping) {
 	oldport := m.ExternalPort()
+
 	log.Debugf("Attempting port map: %s/%d", m.Protocol(), m.InternalPort())
-	newport, err := nat.nat.AddPortMapping(m.Protocol(), m.InternalPort(), "http", MappingDuration)
+	comment := "libp2p"
+	if m.comment != "" {
+		comment = "libp2p-" + comment
```
Comment (geoah) [inline p2p/nat/nat.go]: @Kubuxu erm is there any change you meant `comment = "libp2p-" + m.comment`?

Comment (Kubuxu) [inline p2p/nat/nat.go]: Yes, thanks.

File: p2p/nat/nat.go [hunk 2]
```diff
@@ -333,13 +336,23 @@ func (nat *NAT) NewMapping(maddr ma.Multiaddr) (Mapping, error) {
 
 func (nat *NAT) establishMapping(m *mapping) {
 	oldport := m.ExternalPort()
+
 	log.Debugf("Attempting port map: %s/%d", m.Protocol(), m.InternalPort())
-	newport, err := nat.nat.AddPortMapping(m.Protocol(), m.InternalPort(), "http", MappingDuration)
+	comment := "libp2p"
+	if m.comment != "" {
```
Comment (whyrusleeping) [inline p2p/nat/nat.go]: what if we get back our own comment here? would this end up with "libp2p-libp2p-libp2p-libp2p...." ?

Comment (Kubuxu) [inline p2p/nat/nat.go]: there is `m.comment `and `comment`.
`comment` is temporary variable. I wanted to use `context.WithValue` to pass it but unfortunately goprocess doesn't have something like it.

Comment (whyrusleeping) [inline p2p/nat/nat.go]: Ah, i see. I had assumed the comment field came from a prexisting NAT mapping on a router

**Comment by whyrusleeping** [comment]
@Kubuxu Not sure if we should try to fix it here or not, but we have a race condition in the old code: https://github.com/ipfs/go-ipfs/issues/3205

Can you check and see if a similar issue is present in the new code?

**Comment by whyrusleeping** [comment]
I think the fix is to just lock around any time we use the 'nat' object

**Comment by Kubuxu** [comment]
The code didn't change so the issue still will be there.
Will lock around. Or maybe we should chane it into worker loop?

**Comment by whyrusleeping** [comment]
@Kubuxu hrm... i think for now the easiest thing will be to just lock around it.

**Comment by whyrusleeping** [comment]
Those changes LGTM, but things don't appear to build

**Comment by Kubuxu** [comment]
Hmm, I will check if once more.

**Comment by whyrusleeping** [comment]
LGTM

**Pull request merged**